lib/eventum/auth/class.cas_auth_backend.php

```diff
@@ -0,0 +1,356 @@
+            'external_id' => $remote['uid'],
+        );
+
+        if (!empty($setup['customer_id_attribute'])) {
+            $data['customer_id'] = $remote[$setup['customer_id_attribute']];
+        }
+        if (!empty($setup['contact_id_attribute'])) {
+            $data['contact_id'] = $remote[$setup['contact_id_attribute']];
+        }
+
+        // if local user found, update it and return usr id
+        if ($usr_id) {
+            // do not reset user password, it maybe be set locally before this
+            unset($data['password']);
+
+            // perspective what is main address and what is alias may be different in ldap and in eventum
+            $emails = array($remote['mail']);
+            $email = User::getEmail($usr_id);
+
+            if (($key = array_search($email, $emails)) !== false) {
+                unset($emails[$key]);
+                $data['email'] = $email;
+            } else {
+                if (count($emails) < 1) {
+                    throw new AuthException('E-mail is required');
+                }
+                // just use first email
+                $data['email'] = array_shift($emails);
+            }
+
+            // do not clear full name if for some reason it is empty
+            if (empty($data['full_name'])) {
+                unset($data['full_name']);
+            }
+
+            $update = User::update($usr_id, $data, false);
+            if ($update > 0) {
+                $this->updateAliases($usr_id, $emails);
+            }
+
+            return $usr_id;
+        } else {
+            // create new local user
+            $setup = self::loadSetup();
+            if ($setup['create_users'] == false) {
+                throw new AuthException('User does not exist and will not be created.');
+            }
+            $data['role'] = $setup['default_role'];
+
+            $emails = array($remote['mail']);
+            if (count($emails) < 1) {
+                throw new AuthException('E-mail is required');
+            }
+            $data['email'] = array_shift($emails);
+
+            if (!empty($data['customer_id']) && !empty($data['contact_id'])) {
+                foreach ($data['role'] as $prj_id => $role) {
+                    if ($role > 0) {
+                        $data['role'][$prj_id] = User::ROLE_CUSTOMER;
+                    }
+                }
+            }
+            $usr_id = User::insert($data);
+            if ($usr_id > 0 && $emails) {
+                $this->updateAliases($usr_id, $emails);
+            }
+        }
+
+        return $usr_id;
+    }
+
+    private function updateAliases($usr_id, $aliases)
+    {
+        foreach ($aliases as $alias) {
+            User::addAlias($usr_id, $alias);
+        }
+    }
+
+
+    /**
+     * With CAS we cannot do a simple password check like this. This will prevent the CLI from working
+     * so at some point in the future we need to find a solution.
+     *
```

Comment:
Agreed that this would be the best way to go. When I implement CAS internally I will either have to setup auth keys or some workaround for this.